File: notification/notification_api.go
package main

import (
	"notification/config"

	"github.com/gin-contrib/cors"
	"github.com/gin-gonic/gin"
	"github.com/sirupsen/logrus"

	"net/http"
	"net/smtp"
	"os"
)

var (
	configFile = os.Getenv("CONFIG_FILE")
	auth       smtp.Auth
)

// EmployeeInfo struct will be the data structure for employee's information
type EmployeeInfo struct {
	ID            string  `json:"id"`
	Name          string  `json:"name"`
	JobRole       string  `json:"job_role"`
	JoiningDate   string  `json:"joining_date"`
	Address       string  `json:"address"`
	City          string  `json:"city"`
	EmailID       string  `json:"email_id"`
	AnnualPackage float64 `json:"annual_package"`
	PhoneNumber   string  `json:"phone_number"`
}

func main() {
	conf, err := config.ParseFile(configFile)
	logrus.SetFormatter(&logrus.JSONFormatter{})
	if err != nil {
		logrus.Errorf("Unable to parse configuration file for notification: %v", err)
	}
	logrus.Infof("Running notification in webserver mode")
	logrus.Infof("notification is listening on port: %v", conf.Notification.APIPort)
	logrus.Infof("Endpoint is available now - http://0.0.0.0:%v/create", conf.Notification.APIPort)
	router := gin.Default()
	config := cors.DefaultConfig()
	config.AllowOrigins = []string{"*"}
	router.Use(cors.New(config))
	router.POST("/notification/send", sendNotification)
	router.GET("/notification/healthz", healthCheck)
	router.Run(":" + conf.Notification.APIPort)
}

func sendNotification(c *gin.Context) {
	conf, err := config.ParseFile(configFile)
	logrus.SetFormatter(&logrus.JSONFormatter{})
	if err != nil {
		logrus.Errorf("Unable to parse configuration file for notification: %v", err)
	}
	var request EmployeeInfo
	if err := c.BindJSON(&request); err != nil {
		errorResponse(c, http.StatusBadRequest, "Malformed request body")
		logrus.Errorf("Error parsing the request body in JSON: %v", err)
		return
	}
	auth = smtp.PlainAuth("", conf.SMTP.Username, conf.SMTP.Password, conf.SMTP.SMTPServer)
	to := request.EmailID

	msg := "From: " + conf.SMTP.From + "\n" +
		"To: " + to + "\n" +
		"Subject: Hello there\n\n" +
		"Your user is added in Opstree Employee Service"
	err = smtp.SendMail(conf.SMTP.SMTPServer+":"+conf.SMTP.SMTPPort, auth, conf.SMTP.From, []string{to}, []byte(msg))
	if err != nil {
		logrus.Errorf("Unable to send mail %v", err)
		errorResponse(c, http.StatusBadRequest, "Unable to send mail")
	}
	c.JSON(http.StatusOK, "Successfully sent notification")
}

func healthCheck(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{
		"message":     "Notification service is running successfully",
		"status_code": 200,
	})
}

func errorResponse(c *gin.Context, code int, err string) {
	c.JSON(code, gin.H{
		"error": err,
	})
}

File: notification/notification_api.py
#!/usr/bin/python3
#pylint: disable = invalid-name, broad-except
"""
A notification application which runs on scheduled basis and send the information to users.
Author:- Opstree Solutions
"""

import argparse
import os
import sys
import logging
import time
import emails
import config_with_yaml as config
from elasticsearch import Elasticsearch
import schedule

CONFIG_FILE = os.environ.get("CONFIG_FILE")
FORMATTER = logging.Formatter("%(asctime)s — %(name)s — %(levelname)s — %(message)s")

def init_logger():
    """Function to initialize logger"""
    console_handler = logging.StreamHandler(sys.stdout)
    console_handler.setFormatter(FORMATTER)
    return console_handler

def get_logger():
    """Function to get logger"""
    logger = logging.getLogger("notification-service")
    logger.setLevel(logging.DEBUG)
    logger.addHandler(init_logger())
    return logger

def read_configuration():
    """This function will read the complete configuration file"""
    logger = get_logger()
    try:
        cfg = config.load(CONFIG_FILE)
        return cfg
    except Exception as e:
        logger.error("Not able to parse the configuration file: %s", e)

def send_mail(email_id):
    """function which will send mail to user"""
    logger = get_logger()
    config_content = read_configuration()
    try:
        message = emails.html(
            html="<strong>Your salary slip is generated please check</strong>",
            subject="Salary Slip",
            mail_from=config_content.getProperty("smtp.from"),
        )

        message.send(
            to=email_id,
            smtp={
                "host": config_content.getProperty("smtp.smtp_server"),
                "port": config_content.getProperty("smtp.smtp_port"),
                "timeout": 5,
                "user": config_content.getProperty("smtp.username"),
                "password": config_content.getProperty("smtp.password"),
                "tls": True,
            },
        )
    except Exception as e:
        logger.error("Not able to send the mail: %s", e)

def send_mail_to_all_users():
    """This function will fetch user information from elasticsearch"""
    logger = get_logger()
    config_content = read_configuration()

    try:
        es_client = Elasticsearch(
            [config_content.getProperty("elasticsearch.host")],
            http_auth=(config_content.getProperty("elasticsearch.username"), \
                config_content.getProperty("elasticsearch.password")),
            scheme="http",
            port=config_content.getProperty("elasticsearch.port"),
        )

        result = es_client.search(
            index="employee-management",
            body={
                "query": {
                    "match_all": {}
                }
            }
        )

        for data in result["hits"]["hits"]:
            send_mail(data["_source"]["email_id"])

    except Exception as e:
        logger.error("Error while executing elasticsearch query: %s", e)

def schedule_operation():
    """This function will gets executed for a scheduled interval"""
    logger = get_logger()
    schedule.every().hour.do(send_mail_to_all_users)
    while True:
        logger.info("Waiting for scheduled time to come, I am sure it will come")
        schedule.run_pending()
        time.sleep(1)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-m", "--mode", \
        help="Mode in which application will run, options - scheduled and external", \
        default="scheduled")
    args = parser.parse_args()

    if args.mode == "scheduled":
        schedule_operation()
    else:
        send_mail_to_all_users()
